044-Stable-identities-for-Kafka-Connect-worker-nodes.md

```diff
@@ -0,0 +1,176 @@
+Since the worker identity is based on the Pod address, it means that we need to give the pods stable addresses.
+
+To achieve that, we should migrate the Kafka Connect clusters from Kubernetes Deployment resource to the StrimziPodSet resources.
+The Pods created by the StrimziPodSet will use stable pod names instead of the randomized names used for the Pods managed by Kubernetes Deployments.
+The pods will be named `<cluster-name>-connect-<index>` where the `<cluster-name>` is the name of the custom resource and the `<index>` will be a sequential index number starting with 0.
+For example, the pod names for Kafka Connect cluster created by `KafkaConnect` custom resource named `my-connect` with 3 replicas will be `my-connect-connect-0`, `my-connect-connect-1`, and `my-connect-connect-2`
+This pattern follows the way the pods are currently named as well.
+
+In addition to the StrimziPodSet resource used to manage the Pods, Strimzi will create a headless Kubernetes Service to give these pods a stable DNS names.
+The headless service will be named `<cluster-name>-connect`.
+For example `my-connect-connect`.
+The existing `ClusterIP` type service names `<cluster-name>-connect-api` will remain unchanged and will round-robin across all the Connect worker pods.
+The combination of the stable Pod names and the headless service will mean that each Kafka Connect worker pod will have a stable DNS name in the format `<cluster-name>-connect-<index>.<cluster-name>-connect.<namespace>.svc`.
+For example `my-connect-connect-0.my-connect-connect.myproject.svc`.
+This DNS name will be used as the advertised address for the worker and give it the stable identity.
+
+### Rolling updates
+
+Since the Pods will now have stable names, the rolling update will always first delete the old pod before creating the new pod.
+There is no way to first create the new pod before deleting the old Pod.
+
+The CO will decide if a rolling update is needed based on the differences between the existing Pods and the desired Pods.
+Today, this is done by the Kubernetes Deployment Controller.
+With StrimziPodSets, this would need to be done by the CO directly.
+The CO will:
+* Loop through all pods / replicas starting with the lowest index number and ending with the highest index number
+* For each pod, it will:
+    * Stop the old pod
+    * Wait for the new Pod to be started (this will be done by the StrimziPodSetController)
+    * Once the new Pod is started, it will wait until it becomes ready before moving to the next replica
+
+### Scaling up or down
+
+When scaling-up, Strimzi will start new Connect pods with the lowest available (unused) index numbers.
+When scaling-down, the Pods will removed starting with the highest used index numbers.
+Since all Pods will share the same configuration, there is no need to support things such as scaling own some pods from the middle or beginning of the sequence.
```

Comment:
```suggestion
Since all Pods will share the same configuration, there is no need to support things such as scaling pods from the middle or beginning of the sequence.
```